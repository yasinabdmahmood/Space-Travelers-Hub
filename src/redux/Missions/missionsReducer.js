import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const baseUrl = 'https://api.spacexdata.com/v3/missions';

export const getMissions = createAsyncThunk(
  'getMissions/',
  async () => {
    try {
      return axios.get(baseUrl);
    } catch (error) {
      return error;
    }
  },
);

const initialState = [];

export const MissionSlice = createSlice({
  name: 'missions',
  initialState,
  reducers: {
    toggleJoining(state, action) {
      const newState = state.map((el) => (
        el.id === action.payload
          ? { ...el, joined: !el.joined } : el));
      return newState;
    },

  },
  extraReducers: (builder) => {
    builder.addCase(getMissions.fulfilled, (state, action) => {
      const missions = action.payload.data.map((el) => ({
        id: el.mission_id,
        mission_name: el.mission_name,
        description: el.description,
        joined: false,
      }));
      return missions;
    });
  },
});
export const { toggleJoining } = MissionSlice.actions;
export default MissionSlice.reducer;
