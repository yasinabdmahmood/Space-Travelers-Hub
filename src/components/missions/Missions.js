import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { toggleJoining } from '../../redux/Missions/missionsReducer';
import style from './missions.module.css';

function Missions() {
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  return (
    <>
      <table className={style.table} data-testid="list-of-missions">
        <thead className={style.thead}>
          <tr className={style.header}>
            <th>Mission</th>
            <th>Description</th>
            <th>Status</th>
            <th> </th>
          </tr>
        </thead>
        {missions.map((el) => (
          <tbody key={el.id} className={style.item}>
            <td className={style.row}>{el.mission_name}</td>
            <p className={style.description}>{el.description}</p>
            <td>
              <button type="submit" className={el.joined ? style.inactiveMember : style.activeMember}>
                {
                el.joined ? 'NOT A MEMBER' : 'Active Member'
              }
              </button>
            </td>
            <td>
              <button type="submit" onClick={() => dispatch(toggleJoining(el.id))} className={el.joined ? style.leave : style.join}>
                { el.joined ? 'Leave Mission' : 'Join Mission'}
              </button>
            </td>
          </tbody>
        ))}
      </table>
    </>
  );
}

export default Missions;
