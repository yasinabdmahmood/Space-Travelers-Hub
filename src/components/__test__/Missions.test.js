import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import Missions from '../missions/Missions';
import store from '../../redux/configureStore';
import '@testing-library/jest-dom';

it('Check if the component has changed', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <Missions />
    </Provider>,

  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('the component should render a div ', () => {
  render(
    <Provider store={store}>
      <Missions />
    </Provider>,
  );

  const mission = screen.getByTestId('list-of-missions');

  expect(mission).toBeInTheDocument();
});
