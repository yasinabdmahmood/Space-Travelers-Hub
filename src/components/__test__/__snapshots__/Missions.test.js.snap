// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if the component has changed 1`] = `
<table
  className="table"
  data-testid="list-of-missions"
>
  <thead
    className="thead"
  >
    <tr
      className="header"
    >
      <th>
        Mission
      </th>
      <th>
        Description
      </th>
      <th>
        Status
      </th>
      <th>
         
      </th>
    </tr>
  </thead>
</table>
`;
